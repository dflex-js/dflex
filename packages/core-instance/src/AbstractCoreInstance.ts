import type { AbstractCoreInterface, AbstractCoreInput } from "./types";

class AbstractCoreInstance implements AbstractCoreInterface {
  ref!: HTMLElement | null;

  id: string;

  translateY?: number;

  translateX?: number;

  isInitialized: boolean;

  isPaused: boolean;

  /**
   * Creates an instance of AbstractCoreInstance.
   */
  constructor({
    ref,
    id,
    isPaused = false,
    isInitialized = true,
  }: AbstractCoreInput) {
    this.id = id;

    this.isInitialized = isInitialized;
    this.isPaused = isPaused;

    if (this.isInitialized) {
      this.attach(ref);
    } else {
      this.ref = null;
    }

    if (!this.isPaused) {
      this.initTranslate();
    }
  }

  hasValidRef() {
    return this.isInitialized && this.ref !== null && this.ref.isConnected;
  }

  attach(incomingRef: HTMLElement | null) {
    // Cleanup.
    this.ref = null;

    if (!incomingRef) {
      // TODO: Is this always running when document is valid? Or it's buggy?
      const ref = document.getElementById(this.id);
      if (!ref) {
        throw new Error(`DFlex: Element with ID: ${this.id} is not found.`);
      }
    } else if (incomingRef.nodeType !== Node.ELEMENT_NODE) {
      throw new Error(
        `DFlex: Invalid HTMLElement: ${incomingRef} is passed to registry.`
      );
    }

    this.ref = incomingRef;
    this.isInitialized = true;
  }

  detach() {
    this.isInitialized = false;
    this.isPaused = true;
    this.ref = null;
  }

  initTranslate() {
    /**
     * Since element render once and being transformed later we keep the data
     * stored to navigate correctly.
     *
     * If it's already initiated we don't need to do it again.
     * Reason: You may detach ref set flag to false and then attach it again. Do
     * you want to start from zero or maintain the last position.
     *
     * Continuity is fundamental in DFlex, please keep that in your mind.
     */
    if (typeof this.translateY !== "number") this.translateY = 0;
    if (typeof this.translateX !== "number") this.translateX = 0;

    this.isPaused = false;
  }
}

export default AbstractCoreInstance;
