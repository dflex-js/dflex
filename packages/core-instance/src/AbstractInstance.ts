/* eslint-disable max-classes-per-file */
import { AxesCoordinates } from "@dflex/utils";

import type {
  AbstractInterface,
  AbstractInput,
  AbstractOpts,
  AllowedAttributes,
} from "./types";

class AbstractInstance implements AbstractInterface {
  ref!: HTMLElement | null;

  id: string;

  translate!: AxesCoordinates;

  isInitialized!: boolean;

  isPaused!: boolean;

  hasAttribute!: {
    // eslint-disable-next-line no-unused-vars
    [key in AllowedAttributes]: boolean;
  };

  constructor({ ref, id }: AbstractInput, opts: AbstractOpts) {
    this.id = id;

    if (opts.isInitialized) {
      this.attach(ref || null);

      this.isPaused = opts.isPaused;

      if (!this.isPaused) {
        this.initTranslate();
      }

      return;
    }

    this.isInitialized = false;
    this.ref = null;
    this.isPaused = true;
  }

  /**
   * Attach element DOM node to the instance.
   *
   * @param incomingRef
   */
  attach(incomingRef: HTMLElement | null) {
    // Cleanup.
    this.ref = null;

    if (!incomingRef) {
      const ref = document.getElementById(this.id);

      if (!ref) {
        throw new Error(`DFlex: Element with ID: ${this.id} is not found.`);
      }
    } else if (incomingRef.nodeType !== Node.ELEMENT_NODE) {
      throw new Error(
        `DFlex: Invalid HTMLElement: ${incomingRef} is passed to registry.`
      );
    }

    this.ref = incomingRef;
    this.isInitialized = true;
  }

  /**
   * Detach element DOM node from the instance.
   */
  detach() {
    this.isInitialized = false;
    this.ref = null;
  }

  transform(x: number, y: number) {
    this.ref!.style.transform = `translate3d(${x}px,${y}px, 0)`;
  }

  /**
   * Initialize the translate AxesCoordinates as part of abstract instance and
   * necessary for darg only movement.
   */
  initTranslate() {
    if (!this.translate) {
      this.translate = new AxesCoordinates(0, 0);

      // @ts-expect-error - Just for initialization.
      this.hasAttribute = {};
    }

    this.isPaused = false;
  }

  updateDatasetIndex(i: number) {
    this.ref!.dataset.index = `${i}`;
  }

  setAttribute(key: AllowedAttributes, value: string) {
    if (this.hasAttribute[key] === true) {
      // console.log(`DFlex: Attribute ${key} is already set.`);

      return;
    }

    this.ref!.setAttribute(key, value);
    this.hasAttribute[key] = true;
  }

  removeAttribute(key: AllowedAttributes) {
    if (this.hasAttribute[key] === true) {
      this.ref!.removeAttribute(key);
      this.hasAttribute[key] = false;
    }
  }

  clearAttributes() {
    (Object.keys(this.hasAttribute) as AllowedAttributes[]).forEach((key) => {
      if (this.hasAttribute[key]) {
        this.removeAttribute(key);
      }
    });

    // @ts-expect-error.
    this.hasAttribute = {};
  }
}

export default AbstractInstance;
