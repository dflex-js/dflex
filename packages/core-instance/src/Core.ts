/* eslint-disable no-param-reassign */

import { PointNum } from "@dflex/utils";
import type {
  RectDimensions,
  Direction,
  Axes,
  IPointNum,
  IPointAxes,
} from "@dflex/utils";

import Abstract from "./Abstract";

import type {
  Keys,
  Order,
  TransitionHistory,
  CoreInput,
  AbstractOpts,
  ICore,
} from "./types";

class Core extends Abstract implements ICore {
  offset!: RectDimensions;

  currentPosition!: IPointNum;

  grid!: IPointNum;

  order: Order;

  keys: Keys;

  depth: number;

  isVisible: boolean;

  hasToTransform!: boolean;

  animatedFrame: number | null;

  #translateHistory?: TransitionHistory[];

  constructor(eleWithPointer: CoreInput, opts: AbstractOpts) {
    const { order, keys, depth, scrollX, scrollY, ...element } = eleWithPointer;

    super(element, opts);

    this.order = order;
    this.keys = keys;
    this.depth = depth;

    this.isVisible = this.isInitialized && !this.isPaused;

    if (this.isInitialized) {
      this.setDataset("index", this.order.self);
    }

    if (!this.isPaused) {
      this.#initIndicators(scrollX, scrollY);
    }

    this.animatedFrame = null;
  }

  #initIndicators(scrollX: number, scrollY: number) {
    const { height, width, left, top } = this.ref!.getBoundingClientRect();

    /**
     * Element offset stored once without being triggered to re-calculate.
     * Instead, using currentOffset object as indicator to current
     * offset/position. This offset, is the init-offset.
     */
    this.offset = {
      height,
      width,

      left: left + scrollX,
      top: top + scrollY,
    };

    this.currentPosition = new PointNum(this.offset.left, this.offset.top);

    /**
     * Initializing grid comes later when the siblings boundaries are
     * initialized and the element is visible.
     */
    this.grid = new PointNum(0, 0);

    this.hasToTransform = false;
  }

  #updateCurrentIndicators(space: IPointAxes) {
    this.translate.increase(space);

    const { left, top } = this.offset!;

    /**
     * This offset related directly to translate Y and Y. It's isolated from
     * element current offset and effects only top and left.
     */
    this.currentPosition!.setAxes(
      left + this.translate.x,
      top + this.translate.y
    );

    if (!this.isVisible) this.hasToTransform = true;
  }

  resume(scrollX: number, scrollY: number) {
    if (!this.isInitialized) this.attach(null);

    this.initTranslate();

    this.#initIndicators(scrollX, scrollY);
  }

  changeVisibility(isVisible: boolean) {
    if (isVisible === this.isVisible) return;

    this.isVisible = isVisible;

    if (this.hasToTransform && this.isVisible) {
      this.transformElm();
      this.hasToTransform = false;
    }
  }

  transformElm() {
    if (this.animatedFrame !== null) {
      window.cancelAnimationFrame(this.animatedFrame);
    }

    this.animatedFrame = window.requestAnimationFrame(() => {
      this.transform(this.translate.x, this.translate.y);
      this.animatedFrame = null;
    });
  }

  #updateOrderIndexing(i: number) {
    const { self: oldIndex } = this.order;

    const newIndex = oldIndex + i;

    this.order.self = newIndex;

    this.setDataset("index", newIndex);

    return { oldIndex, newIndex };
  }

  assignNewPosition(branchIDsOrder: string[], newIndex: number) {
    if (newIndex < 0 || newIndex > branchIDsOrder.length - 1) {
      if (process.env.NODE_ENV !== "production") {
        // eslint-disable-next-line no-console
        console.error(
          `Illegal Attempt: Received an index:${newIndex} on siblings list:${
            branchIDsOrder.length - 1
          }.\n`
        );
      }

      return;
    }

    if (branchIDsOrder[newIndex].length > 0) {
      if (process.env.NODE_ENV !== "production") {
        // eslint-disable-next-line no-console
        console.error(
          "Illegal Attempt: Colliding in positions.\n",
          `Element id: ${this.id}\n`,
          `Siblings list: ${branchIDsOrder}\n`
        );
      }

      return;
    }

    branchIDsOrder[newIndex] = this.id;
  }

  #leaveToNewPosition(
    branchIDsOrder: string[],
    newIndex: number,
    oldIndex: number,
    siblingsEmptyElmIndex: number
  ) {
    if (siblingsEmptyElmIndex >= 0 && siblingsEmptyElmIndex !== newIndex) {
      if (process.env.NODE_ENV !== "production") {
        // eslint-disable-next-line no-console
        console.error(
          `Illegal Attempt: More than one element have left the siblings list.\n`,
          `Element id: ${this.id} - index: ${oldIndex}\n`,
          `Siblings list: ${branchIDsOrder}\n`
        );
      }

      return siblingsEmptyElmIndex;
    }

    branchIDsOrder[oldIndex] = "";

    branchIDsOrder[newIndex] = this.id;

    return siblingsEmptyElmIndex;
  }

  /**
   *  Set a new translate position and store the old one.
   */
  #seTranslate(
    elmSpace: IPointAxes,
    axis: Axes,
    operationID?: string,
    isForceTransform = false
  ) {
    if (operationID) {
      const elmAxesHistory: TransitionHistory = {
        ID: operationID,
        axis,
        translate: { x: this.translate.x, y: this.translate.y },
      };

      if (!Array.isArray(this.#translateHistory)) {
        this.#translateHistory = [elmAxesHistory];
      } else {
        this.#translateHistory.push(elmAxesHistory);
      }
    }

    this.#updateCurrentIndicators(elmSpace);

    if (!isForceTransform && !this.isVisible) {
      this.hasToTransform = true;

      return;
    }

    this.transformElm();

    this.hasToTransform = false;
  }

  /**
   *
   * @param iDsInOrder -
   * @param direction - (+1/-1)
   * @param elmSpace - space between dragged and the immediate next element.
   * @param operationID - A unique ID used to store translate history
   */
  setPosition(
    iDsInOrder: string[],
    direction: Direction,
    elmSpace: IPointNum,
    operationID: string,
    siblingsEmptyElmIndex: number,
    axis: Axes
  ) {
    const numberOfPassedElm = 1;

    /**
     * effectedElemDirection decides the direction of the element, negative or positive.
     * If the element is dragged to the left, the effectedElemDirection is -1.
     */
    if (axis === "z") {
      elmSpace.multiplyAll(direction);
    } else {
      elmSpace[axis] *= direction;
    }

    this.#seTranslate(elmSpace, axis, operationID);

    const { oldIndex, newIndex } = this.#updateOrderIndexing(
      direction * numberOfPassedElm
    );

    if (axis === "z") {
      const inc = direction * numberOfPassedElm;
      this.grid.increase({ x: inc, y: inc });
    } else {
      this.grid[axis] += direction * numberOfPassedElm;
      if (process.env.NODE_ENV !== "production") {
        this.setDataset(
          `grid${axis.toUpperCase() as "X" | "Y"}`,
          this.grid[axis]
        );
      }
    }

    const newStatusSiblingsHasEmptyElm = this.#leaveToNewPosition(
      iDsInOrder,
      newIndex,
      oldIndex,
      siblingsEmptyElmIndex
    );

    return newStatusSiblingsHasEmptyElm;
  }

  /**
   * Roll back element position.
   *
   * @param operationID
   * @param isForceTransform
   */
  rollBack(operationID: string, isForceTransform: boolean) {
    if (
      !Array.isArray(this.#translateHistory) ||
      this.#translateHistory.length === 0 ||
      this.#translateHistory[this.#translateHistory.length - 1].ID !==
        operationID
    ) {
      return;
    }

    const lastMovement = this.#translateHistory.pop()!;

    const { translate: preTranslate, axis } = lastMovement;

    const elmSpace = {
      x: preTranslate.x - this.translate.x,
      y: preTranslate.y - this.translate.y,
    };

    let increment = 0;

    if (axis === "z") {
      increment = elmSpace.x > 0 || elmSpace.y > 0 ? 1 : -1;

      this.grid.increase({ x: increment, y: increment });
    } else {
      increment = elmSpace[axis] > 0 ? 1 : -1;

      this.grid[axis] += increment;

      if (process.env.NODE_ENV !== "production") {
        this.setDataset(
          `grid${axis.toUpperCase() as "X" | "Y"}`,
          this.grid[axis]
        );
      }
    }

    // Don't update UI if it's zero and wasn't transformed.
    this.#seTranslate(elmSpace, axis, undefined, isForceTransform);

    this.#updateOrderIndexing(increment);

    this.rollBack(operationID, isForceTransform);
  }
}

export default Core;
