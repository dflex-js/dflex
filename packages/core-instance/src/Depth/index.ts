export type { IDepth } from "./types";

export { default as Depth } from "./Depth";
