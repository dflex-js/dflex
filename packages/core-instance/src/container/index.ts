export type { IContainer } from "./types";

export { default as Container } from "./Container";
