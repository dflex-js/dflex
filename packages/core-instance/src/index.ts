export type {
  INodeAbstract as IAbstract,
  CoreInput,
  TransitionHistory,
  INode,
} from "./node";

export { Node } from "./node";
