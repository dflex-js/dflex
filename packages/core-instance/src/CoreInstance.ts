/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable no-param-reassign */

import AbstractCoreInstance from "./AbstractCoreInstance";

import type {
  Keys,
  Order,
  ElmWIthPointer,
  CoreInstanceInterface,
  Offset,
  TransitionHistory,
} from "./types";

/**
 * Why storing index here? when it's already sorted in order?
 *
 * Each element has index elements to element's position in list. It allows us
 * to avoid looping in idsTreeOrder to know each element position.
 *
 * elem-id = id1
 * iDsInOrder[id3, id1, id2]
 * How do we get elem-id from iDsInOrder without loop? We don't want to loop
 * twice once to know the qualified element to transform and the second to
 * figure out element position.
 *
 * So, iDsInOrder[index] is our position in list. And, to update to
 * new position: iDsInOrder[new-index] = elem-id.
 *
 *
 * Why storing prev-index?
 *
 * Currently, there's no undo history. But, still need one-step back. Why?
 * Because if dragged is out and release, it should go back to its
 * prevIndex aka selfIndex before transformation.
 *
 * why Storing parent-index?
 *
 * To connect element with parents by knowing their locations.
 */

class CoreInstance
  extends AbstractCoreInstance
  implements CoreInstanceInterface
{
  offset!: Offset;

  /** Store history of Y-transition according to unique ID. */
  prevTranslateY: TransitionHistory;

  currentTop: number;

  currentLeft: number;

  scrollWidth?: number;

  scrollHeight?: number;

  clientWidth?: number;

  clientHeight?: number;

  order: Order;

  keys: Keys;

  constructor(elementWithPointer: ElmWIthPointer) {
    const { order, keys, ...element } = elementWithPointer;

    super(element);

    this.prevTranslateY = [];

    /**
     * Used for dragged, storing temporary top, left new positions during the transition.
     */
    this.currentTop = 0;
    this.currentLeft = 0;

    if (this.ref) {
      this.initIndicators();
    }

    this.order = order;
    this.keys = keys;

    if (this.keys.chK !== null) this.getScroll();
  }

  /**
   * Initializes the element offset only when it's called. Since it is sorting
   * different numbers related to transformation we don't need to invoke for
   * idle element because it's costly.
   *
   * So, basically any working element in DnD should be initiated first.
   */
  private initIndicators() {
    const { height, width, left, top } = this.ref.getBoundingClientRect();

    /**
     * Element offset stored once without being triggered to re-calculate.
     * Instead, using currentOffset object as indicator to current
     * offset/position. This offset, is the init-offset.
     */
    this.offset = {
      height,
      width,

      left,
      top,
    };

    this.currentTop = top;
    this.currentLeft = left;
  }

  private getScroll() {
    const { scrollWidth, scrollHeight, clientWidth, clientHeight } = this.ref;

    this.scrollWidth = scrollWidth;
    this.clientWidth = clientWidth;
    this.scrollHeight = scrollHeight;
    this.clientHeight = clientHeight;
  }

  getOffset() {
    return {
      height: this.offset.height,
      width: this.offset.width,

      left: this.currentLeft,
      top: this.currentTop,
    };
  }

  private updateCurrentIndicators(topSpace: number, leftSpace: number) {
    this.translateY += topSpace;
    this.translateX += leftSpace;

    const { left, top } = this.offset;

    /**
     * This offset related directly to translate Y and Y. It's isolated from
     * element current offset and effects only top and left.
     */
    this.currentTop = top + this.translateY;
    this.currentLeft = left + this.translateX;
  }

  transformElm() {
    this.ref.style.transform = `translate(${this.translateX}px,${this.translateY}px)`;
  }

  /**
   *  Update element index in order  branch
   *
   * @param i - index
   */
  private updateOrderIndexing(i: number) {
    const { self: oldIndex } = this.order;

    const newIndex = oldIndex + i;

    this.order.self = newIndex;

    return { oldIndex, newIndex };
  }

  assignNewPosition(
    branchIDsOrder: string[],
    newIndex: number,
    oldIndex?: number
  ) {
    // TODO: Add this to confusion mode:newIndex < 0)
    branchIDsOrder[newIndex] = this.id;
    if (oldIndex !== undefined) branchIDsOrder[oldIndex] = "";
  }

  /**
   *  Set a new translate position and store the old one.
   *
   * @param topSpace -
   * @param operationID  - Only if moving to a new position.
   */
  private seTranslate(topSpace: number, operationID?: string) {
    if (operationID) {
      this.prevTranslateY.push({
        ID: operationID,
        translateY: this.translateY,
      });
    }

    this.updateCurrentIndicators(topSpace, 0);
    this.transformElm();
  }

  /**
   * Sets new vertical position. Which includes, TranslateY and OffsetTop. By assigning the
   * new calculated value by +/- new difference.
   *
   * Note: Why we don't need setXPosition?
   * Because, elements always move in the same list container, the only one who's migrated to
   * another is dragged.
   *
   * Note: isShuffle is flag made for updating last element in array
   * which is dragged. Normally, update element position and clear its previous
   * position but when updating last element the array is ready and done we need
   * to update one position only so don't clear previous.
   *
   * @param iDsInOrder -
   * @param sign - (+1/-1)
   * @param topSpace - space between dragged and the immediate next element.
   * @param operationID - A unique ID used to store translate history
   * @param vIncrement - the number of passed elements.
   * @param isShuffle -
   */
  setYPosition(
    iDsInOrder: string[],
    sign: number,
    topSpace: number,
    operationID: string,
    vIncrement = 1,
    isShuffle = true
  ) {
    this.seTranslate(sign * topSpace, operationID);

    const { oldIndex, newIndex } = this.updateOrderIndexing(sign * vIncrement);

    this.assignNewPosition(
      iDsInOrder,
      newIndex,
      isShuffle ? oldIndex : undefined
    );
  }

  /**
   * Roll back element position vertically(y).
   *
   * @param operationID -
   */
  rollYBack(operationID: string) {
    if (
      this.prevTranslateY.length === 0 ||
      this.prevTranslateY[this.prevTranslateY.length - 1].ID !== operationID
    ) {
      return;
    }
    // @ts-ignore
    const { translateY } = this.prevTranslateY.pop();

    const topSpace = translateY - this.translateY;

    const increment = topSpace > 0 ? 1 : -1;

    this.seTranslate(topSpace);
    this.updateOrderIndexing(increment);
  }
}

export default CoreInstance;
