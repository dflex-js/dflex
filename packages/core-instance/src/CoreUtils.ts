import type { Axis } from "@dflex/utils";
import Core from "./Core";
import type { ICoreUtils } from "./types";

class CoreUtils extends Core implements ICoreUtils {
  isPositionedUnder(elmY: number) {
    return elmY < this.currentPosition.y;
  }

  isPositionedLeft(elmX: number) {
    return elmX < this.currentPosition.x;
  }

  hasSamePosition(elm: this, axis: Axis) {
    return this.currentPosition[axis] === elm.currentPosition[axis];
  }

  getRectBottom() {
    return this.currentPosition.y + this.offset.height;
  }

  getRectRight() {
    return this.currentPosition.x + this.offset.width;
  }
}

export default CoreUtils;
