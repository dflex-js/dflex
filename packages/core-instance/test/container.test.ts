import { Container } from "../src";

describe("Container", () => {
  let container: Container;

  beforeAll(() => {
    container = new Container();
  });

  it("container is initiated", () => {
    expect(container).toMatchSnapshot();
  });
});
