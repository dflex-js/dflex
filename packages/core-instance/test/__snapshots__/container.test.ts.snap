// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Container is initiated 1`] = `
Container {
  "grid": PointNum {
    "x": 1,
    "y": 1,
  },
}
`;
